### PR TITLE
Chore/511 improve tim model implementation by updating the timeseries data type

### PR DESCRIPTION
**Pull request by MRVermeulenDeltares**

### CODE REVIEW

**Comment by MRVermeulenDeltares** [comment]
Codesmell and duplication in test, what to do with this?

**Comment by veenstrajelmer** [comment]
@MRVermeulenDeltares: Tested and implementation is as expected

File: hydrolib/core/dflowfm/tim/parser.py [hunk 1]
```diff
@@ -10,25 +12,25 @@ class TimParser:
...
+                - "time" is a time as a string.
+                - "data" is data as a list of strings.
 
         Raises:
-            ValueError: If the file contains a comment that is not at the start of the file or
-             if the time series contains a duplicate time entry.
+            ValueError: If the file contains a comment that is not at the start of the file.
         """
 
         comments: List[str] = []
```
Comment (tim-vd-aardweg) [inline hydrolib/core/dflowfm/tim/parser.py:32]: Once again, a nice clean function. Nice! 👍

File: hydrolib/core/dflowfm/tim/serializer.py [hunk 2]
```diff
@@ -27,9 +27,12 @@ def serialize(
 
         Args:
             path (Path): The path to the destination .tim file.
-            data (Dict[str, Any]): The timeseries data to be serialized. The data should be provided as a dictionary with the following keys:
-                - 'comments' (List[str]): A list of comments to be included at the beginning of the file. Each comment should be a string.
-                - 'timeseries' (Dict[float, List[float]]): A dictionary representing the timeseries data, where the key is the time and the value the values at that time.
+            data Dict[List[str], Dict[str, List[str]]]: A dictionary with keys "comments" and "timeseries".\n
```
Comment (tim-vd-aardweg) [inline hydrolib/core/dflowfm/tim/serializer.py:30]:
```suggestion
            data (Dict[List[str], Dict[str, List[str]]]): A dictionary with keys "comments" and "timeseries".\n
```

File: hydrolib/core/dflowfm/tim/serializer.py [hunk 3]
```diff
@@ -18,7 +18,7 @@ class TimSerializer:
     @staticmethod
     def serialize(
         path: Path,
-        data: Dict[str, Any],
+        data: Dict[List[str], Dict[str, List[str]]],
```
[the remark this answers is not in the document]
Comment (tim-vd-aardweg) [reply hydrolib/core/dflowfm/tim/serializer.py:21]: Don't forget to update the docstring! :)

File: hydrolib/core/dflowfm/tim/serializer.py [hunk 4]
```diff
@@ -27,9 +27,12 @@ def serialize(
 
         Args:
             path (Path): The path to the destination .tim file.
-            data (Dict[str, Any]): The timeseries data to be serialized. The data should be provided as a dictionary with the following keys:
-                - 'comments' (List[str]): A list of comments to be included at the beginning of the file. Each comment should be a string.
-                - 'timeseries' (Dict[float, List[float]]): A dictionary representing the timeseries data, where the key is the time and the value the values at that time.
+            data Dict[List[str], Dict[str, List[str]]]: A dictionary with keys "comments" and "timeseries".\n
+            - "comments" is a list of strings representing comments found at the start of the file.\n
```
Comment (tim-vd-aardweg) [inline hydrolib/core/dflowfm/tim/serializer.py:31]: Are the `\n` really needed for formatting?

File: hydrolib/core/dflowfm/tim/serializer.py [hunk 5]
```diff
@@ -65,7 +68,9 @@ def _serialize_timeseries_lines(data, config) -> List[str]:
     @staticmethod
     def _serialize_to_timeseries_block(data, format_float) -> TimeSeriesBlock:
```
Comment (tim-vd-aardweg) [inline hydrolib/core/dflowfm/tim/serializer.py:69]: The typehinting is missing for a couple of the functions in this class!

File: hydrolib/core/dflowfm/tim/serializer.py [hunk 6]
```diff
@@ -65,7 +68,9 @@ def _serialize_timeseries_lines(data, config) -> List[str]:
...
         timeseries_block: TimeSeriesBlock = []
-        for time, row_elements in data["timeseries"].items():
+        for nested_data in data["timeseries"]:
```
Comment (tim-vd-aardweg) [inline hydrolib/core/dflowfm/tim/serializer.py:71]: What is `nested_data`? Can you come up with a word that better describes what is inside so we do not have to use `nested_data`? :D

File: tests/dflowfm/test_tim.py [hunk 7]
```diff
@@ -78,6 +78,28 @@ def test_serialize_data(self, input_data, reference_path):
...
 
 class TestTimModel:
+    triple_data_for_timeseries_floats = [
+        TimRecord(time=10, data=[1.232, 2.343, 3.454]),
+        TimRecord(time=20, data=[4.565, 5.676, 6.787]),
+        TimRecord(time=30, data=[1.5, 2.6, 3.7]),
+    ]
+
+    single_data_for_timeseries_floats = [
+        TimRecord(time=0.0000000, data=[0.0000000]),
```
Comment (tim-vd-aardweg) [inline tests/dflowfm/test_tim.py:88]: Do you really need this many entries? Can't we shorten the testdata to make it more readable? (And to make the file smaller)

File: tests/dflowfm/test_tim.py [hunk 8]
```diff
@@ -186,27 +220,27 @@ def test_validate_data_for_timeseries_throws_exception_for_incorrect_data(
...
-        "50.000000": ["0.0100000"],
-        "60.000000": ["0.0000000"],
-        "70.000000": ["-0.0100000"],
-        "80.000000": ["0.0000000"],
-        "90.000000": ["0.0100000"],
-        "100.000000": ["0.0000000"],
-        "110.000000": ["-0.0100000"],
-        "120.000000": ["0.0000000"],
-    }
+    single_data_for_timeseries = [
```
Comment (tim-vd-aardweg) [inline tests/dflowfm/test_tim.py:229]: Same comment as above: Do we need this much data?

File: tests/data/input/tim/bc_file_is_incorrect.bc [hunk 9]
```diff
@@ -0,0 +1,23 @@
+[forcing]
```
Comment (tim-vd-aardweg) [inline tests/data/input/tim/bc_file_is_incorrect.bc:1]: What is incorrect about this file?

File: tests/dflowfm/test_tim.py [hunk 10]
```diff
@@ -13,30 +13,30 @@
...
-    120.000000: [0.0000000],
-}
+class TestTimSerializer:
+    triple_data_for_timeseries_floats = [
+        {"time": 10, "data": [1.232, 2.343, 3.454]},
+        {"time": 20, "data": [4.565, 5.676, 6.787]},
+        {"time": 30, "data": [1.5, 2.6, 3.7]},
+    ]
 
+    single_data_for_timeseries_floats = [
```
Comment (tim-vd-aardweg) [inline tests/dflowfm/test_tim.py:24]: Do you really need this many entries? Can't we shorten the testdata to make it more readable? (And to make the file smaller)

**Review by tim-vd-aardweg — CHANGES_REQUESTED** [review]

File: hydrolib/core/dflowfm/tim/parser.py [hunk 11]
```diff
@@ -1,5 +1,7 @@
 from pathlib import Path
-from typing import Any, Dict, List, Tuple
+from typing import Dict, List, Tuple
+
+TimRecord = Dict[str, List[str]]
```
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/parser.py:4]: any suggestion for a name for this?

File: tests/data/input/tim/bc_file_is_incorrect.bc (hunk 9 above)
Comment (MRVermeulenDeltares) [reply tests/data/input/tim/bc_file_is_incorrect.bc:1]: the file itself is not incorrect but it is incorrect for tim format, thus better rename bc_file_format_incorrect_for_timmodel?

File: tests/dflowfm/test_tim.py (hunk 10 above)
Comment (MRVermeulenDeltares) [reply tests/dflowfm/test_tim.py:24]: this is the example I got when I started on the tim model

File: tests/dflowfm/test_tim.py (hunk 7 above)
Comment (MRVermeulenDeltares) [reply tests/dflowfm/test_tim.py:88]: this is the example I got when I started on the tim model

File: tests/dflowfm/test_tim.py (hunk 8 above)
Comment (MRVermeulenDeltares) [reply tests/dflowfm/test_tim.py:229]: Same reply as above: this is what I got

File: hydrolib/core/dflowfm/tim/models.py [hunk 12]
```diff
@@ -3,23 +3,35 @@
...
+
+
 class TimModel(ParsableFileModel):
     """Class representing a tim (*.tim) file."""
 
     serializer_config = TimSerializerConfig()
     """TimSerializerConfig: The serialization configuration for the tim file."""
 
-    comments: List[str]
+    comments: List[str] = []
```
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/models.py:30]: also comments added to assert

File: hydrolib/core/dflowfm/tim/models.py [hunk 13]
```diff
@@ -3,23 +3,35 @@
...
+class TimRecord(BaseModel):
+    """Single time record, representing a time and a list of data.
+
+    Attributes:
+        time: time of the time record.
+        data: record of the time recored.
+    """
+
+    time: float
+    data: List[float]
```
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/models.py:21]: lets discuss, this throws validationerror: value_error.missing

File: hydrolib/core/dflowfm/tim/parser.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/parser.py:4]: Changed to TimData

File: hydrolib/core/dflowfm/tim/parser.py [hunk 14]
```diff
@@ -10,25 +12,25 @@ class TimParser:
...
 
         Args:
             filepath (Path): Path to the .tim file to be read.
 
         Returns:
-            Dict[str, Any]: A dictionary with keys "comments" and "timeseries", where "comments"
-                  is a list of strings representing comments found at the start of the file, and
-                  "timeseries" is a dictionary where each key is a time and each value
-                  is a list of strings.
+            Dict[List[str], Dict[str, List[str]]: A dictionary with keys "comments" and "timeseries".\n
```
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/parser.py:22]: don't think so, probably remained when I was playing with getting a dotted list

File: hydrolib/core/dflowfm/tim/serializer.py (hunk 5 above)
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/serializer.py:69]: do you know what type format_float is?

File: hydrolib/core/dflowfm/tim/models.py [hunk 15]
```diff
@@ -3,23 +3,35 @@
...
 
+class TimRecord(BaseModel):
+    """Single time record, representing a time and a list of data.
+
+    Attributes:
+        time: time of the time record.
+        data: record of the time recored.
+    """
+
+    time: float
```
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/models.py:15]: done

File: hydrolib/core/dflowfm/tim/models.py [hunk 16]
```diff
@@ -3,23 +3,35 @@
...
 
-from hydrolib.core.basemodel import ModelSaveSettings, ParsableFileModel
+from hydrolib.core.basemodel import BaseModel, ModelSaveSettings, ParsableFileModel
 
 from .parser import TimParser
 from .serializer import TimSerializer, TimSerializerConfig
 
 
+class TimRecord(BaseModel):
+    """Single time record, representing a time and a list of data.
```
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/models.py]: done

File: hydrolib/core/dflowfm/tim/models.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/models.py:28]: done

File: hydrolib/core/dflowfm/tim/models.py [hunk 17]
```diff
@@ -41,34 +53,45 @@ def _get_parser(cls) -> Callable:
```
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/models.py:51]: done

File: hydrolib/core/dflowfm/tim/models.py [hunk 18]
```diff
@@ -41,34 +53,45 @@ def _get_parser(cls) -> Callable:
...
-            Dict[float, List[float]: Validated timeseries.
+            List[TimRecord]: Validated timeseries.
         """
         if len(v) == 0:
             return v
 
-        timeseries_iterator = iter(v.items())
-        _, columns = next(timeseries_iterator)
-        n_columns = len(columns)
+        cls._raise_error_if_amount_of_columns_differ(v)
```
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/models.py:71]: thx

File: hydrolib/core/dflowfm/tim/parser.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/parser.py]: done

File: hydrolib/core/dflowfm/tim/parser.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/parser.py]: done

[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/parser.py]: done

File: hydrolib/core/dflowfm/tim/parser.py [hunk 19]
```diff
@@ -10,25 +12,25 @@ class TimParser:
...
+            - "comments" is a list of strings representing comments found at the start of the file.\n
+            - "timeseries" is a list of dictionaries with the key as "time" and values as "data".\n
+                - "time" is a time as a string.
+                - "data" is data as a list of strings.
 
         Raises:
-            ValueError: If the file contains a comment that is not at the start of the file or
-             if the time series contains a duplicate time entry.
+            ValueError: If the file contains a comment that is not at the start of the file.
```
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/parser.py:29]: done

File: hydrolib/core/dflowfm/tim/parser.py (hunk 1 above)
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/parser.py:33]: thx!

File: hydrolib/core/dflowfm/tim/serializer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/serializer.py]: done&done

File: hydrolib/core/dflowfm/tim/serializer.py (hunk 4 above)
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/serializer.py]: done

File: hydrolib/core/dflowfm/tim/serializer.py (hunk 6 above)
Comment (MRVermeulenDeltares) [reply hydrolib/core/dflowfm/tim/serializer.py]: done

File: hydrolib/core/dflowfm/tim/models.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (tim-vd-aardweg) [reply hydrolib/core/dflowfm/tim/models.py:15]: Also add the types in the documentation

File: hydrolib/core/dflowfm/tim/models.py [hunk 20]
```diff
@@ -10,15 +10,13 @@
...
-        data: record of the time recored.
-    """
+    """Single tim record, representing a time and a list of data."""
 
     time: float
-    data: List[float]
+    """time: time of the time record."""
+
+    data: List[float] = []
+    """data: record of the time recored"""
```
Comment (tim-vd-aardweg) [inline hydrolib/core/dflowfm/tim/models.py:19]: Typo!

File: hydrolib/core/dflowfm/tim/serializer.py (hunk 5 above)
Comment (tim-vd-aardweg) [reply hydrolib/core/dflowfm/tim/serializer.py]: `Callable[[float], str]`

**Review by tim-vd-aardweg — APPROVED** [review]

**Comment by tim-vd-aardweg** [comment]
Review OK!

**Comment by veenstrajelmer** [comment]
Tested and works as expected. The only thing I notice is that duplicate timesteps are not being validated anymore. We can discuss if that is necessary (maybe a follow up issue)

**Pull request merged**